File: 2209-minimum-white-tiles-after-covering-with-carpets/2209-minimum-white-tiles-after-covering-with-carpets.py
class Solution:
    def minimumWhiteTiles(self, floor: str, numCarpets: int, carpetLen: int) -> int:
        N = len(floor)
        @cache
        def solve(n,c):
            if n<=0:return 0
            # if c<=0: return 1000
            cover = solve(n-carpetLen,c-1) if c else inf
            notcover = int(floor[n-1]) + solve(n-1,c)
            return min(cover,notcover)
            
        return solve(N,numCarpets)

File: 2209-minimum-white-tiles-after-covering-with-carpets/2209-minimum-white-tiles-after-covering-with-carpets.cpp
class Solution {
public:
    int minimumWhiteTiles(string fl, int nc, int cl) {
        int n = fl.size();
        vector<vector<int>>dp(n+1,vector<int>(nc+1,0));
        for(int floorLen=1;floorLen<=n;floorLen++){
            for(int carpHave=0;carpHave<=nc;carpHave++){
                auto cover = carpHave ? dp[max(floorLen-cl,0)][carpHave-1]:INT_MAX;
                auto notcover = fl[floorLen-1]-'0' + dp[floorLen-1][carpHave];
                dp[floorLen][carpHave] = min(cover,notcover);
            }
        }
        
        return dp[n][nc];
        
    }
};